code/modules/library/skill_learning/job_skillchips/traitor.dm

```diff
@@ -0,0 +1,30 @@
+/obj/item/skillchip/chameleon
+	name = "Chameleon skillchip"
+	desc = "A highly advanced Syndicate skillchip that does nothing on its own. It is loaded with the data of every skillchip."
+	skill_name = "Imitate Skillchip"
+	skill_description = "Reacts to the user's thoughts, selecting a skill from a wide database of choices."
+	skill_icon = "microchip"
+	removable = FALSE
+	/// Action for the skillchip selection.
+	var/datum/action/item_action/chameleon/change/skillchip/chameleon_action
+
+/obj/item/skillchip/chameleon/emp_act(severity)
+	. = ..()
+	if(. & EMP_PROTECT_SELF)
+		return
+	chameleon_action.emp_randomise()
+
+/obj/item/skillchip/chameleon/on_apply(mob/living/carbon/user, silent = TRUE)
+	. = ..()
+
+	chameleon_action = new(src)
+	chameleon_action.chameleon_type = /obj/item/skillchip
+	chameleon_action.chameleon_name = "Skillchip"
+	chameleon_action.chameleon_blacklist = typecacheof(list(/obj/item/skillchip, /obj/item/skillchip/chameleon, /obj/item/skillchip/useless_adapter), only_root_path = TRUE)
```

Comment:
`typecacheof` already handles subtypes, you would achieve the same result with `typecacheof(/obj/item/skillchip)`
That said, do you really need a typecache here?
It would be better to make a variable or flag instead, and check on it if you want to populate a whitelist.

Reply:
I'm going to implement a middleground solution and add a flag check alongside this method.

The problem is - I want to exclude certain type paths from the Chameleon item list such as the parent type /obj/item/skillchip.

I can simply add a flag and stop using this blacklist, but then every single child of /obj/item/skillchip HAS to override the flags.

The reason /obj/item/skillchip would have the chameleon incompatibility flag set is not because it should be the default behaviour of skill chips, but because /obj/item/skillchip isn't a REAL skillchip.

Anyone who forgot to clear the flags when inheriting /obj/item/skillchip would have unexpected functionality in their chips. It's effectively an open season for introducing bugs. I suspect this is part of why chameleon items use the blacklist method that they do.

As a result, I'm doing it both ways. I'm going to hardcode "abstract" item paths in chameleon_blacklist however when building the item list, it also checks for the presence of the appropriate flag to exclude the item too.

The end system will be as follows: Real skillchips that want to be chameleon blacklisted can use the flag. Abstract skillchips that exist just as parents and should never actually appear in-game can have their paths added to the blacklist without forcing ALL their children to override thier flags.

Reply:
Testing out an alternative system after talking with AnturK on Discord. Could have found a better way,